python/ray/train/mosaic/_mosaic_utils.py

```diff
@@ -55,14 +60,37 @@ def batch_checkpoint(self, state: State, logger: Logger) -> None:
     def epoch_checkpoint(self, state: State, logger: Logger) -> None:
         del logger  # unused
         self.should_report_fit_end = False
-        session.report(self.data)
-
-        # flush the data
-        self.data = {}
+        self._report(state)
 
     def fit_end(self, state: State, logger: Logger) -> None:
         # report at close in case the trainer stops in the middle of an epoch.
         # this may be double counted with epoch checkpoint.
         del logger  # unused
         if self.should_report_fit_end:
-            session.report(self.data)
+            self._report(state)
+
+    def _report(self, state: State):
+        model_state_dict = state.state_dict()["model"]
```

Comment:
can we have a unit test for this part ?

Also I think just using pytorch's `apply()` will be much simpler here since it recursively walks through all modules and we don't have to iterately keep a while True loop on our own.

I wrote a quick example for appending my own prefix to all modules `state_dict`:

```python
import torch.nn as nn
import torch.nn.functional as F


my_state_dict = {}

def add_prefix(m):
    for key in m.state_dict().keys():
        my_state_dict[key] = "my_prefix." + key

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(1, 20, 5)
        self.conv2 = nn.Conv2d(20, 20, 5)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        return F.relu(self.conv2(x))

    
model = Model()
print(f" Original state dict keys: \n{list(model.state_dict().keys())} \n")

model.apply(add_prefix)

print(f" Recursive added state dict keys: \n{list(my_state_dict.values())} \n")
```

that outputs

```
 Original state dict keys: 
['conv1.weight', 'conv1.bias', 'conv2.weight', 'conv2.bias'] 

 Recursive added state dict keys: 
['my_prefix.weight', 'my_prefix.bias', 'my_prefix.conv1.weight', 'my_prefix.conv1.bias', 'my_prefix.conv2.weight', 'my_prefix.conv2.bias'] 
```

Reply:
```
        # remove module prefixes when loading the model weights
        # two ensure that all possible repetitions of "module." is removed, we use
        # a while loop
        prefix = "module."
        keys = list(model_state_dict.keys())
        for key in keys:
            check_key = key
            while check_key.startswith(prefix):
                newkey = check_key[len(prefix) :]
                model_state_dict[newkey] = model_state_dict.pop(check_key)
                check_key = newkey

```

Would this be a better way to remove all possible additions of "module" prefix?. This is to provide flexibility in the number of wrappers that may have been added in the model state dict. In general, the "module" wrapping is present twice, so to make it deterministic, we can use `torch.nn.modules.utils.consume_prefix_in_state_dict_if_present` twice

Reply:
hmm so iiuc you're dealing with module names potentially as `module.module.conv` and `module.module.module.conv` and only want to keep the `conv` part with all `module` prefix removed regardless of num layers ?  Then a single `apply` won't be sufficient. Again, can we have a unit test for this function with representative module structures you're targeted solving with this snippet first, then we can see if this is the easiest way with concrete model examples